### PR TITLE
Modular CLI

### PR DESCRIPTION
**Pull request by LokiLuciferase**
Making recently introduced functionality (XGB, parameter search...) available via the `argparse` CLI interface would require most subcommands to be able to handle differing (but largely shared among subcommands) sets of options.
This would make the interface either excessively WET, or necessitate high number of confusing conditional statements.

To make our CLI more extendable, modular and testable, I have re-implemented it using the [click](https://www.github.com/pallets/click) framework. 

Looking forward to the package name change, the new CLI is (for now) invoked with `trex`.
The following commands are available:
- `predict`: equivalent to previous command `predict`
- `get-weights`: equivalent to previous command `weights`
- `train`: roughly equivalent to previous command, but takes as additional subcommand the type of classifier to train (`svm` or `xgb`). Allows passing of a training parameters file in JSON format using the flag `--params_file`.
- `cv`: roughly equivalent to previous command `crossvalidate`, same extras as `train`. Allows finding and saving of optimal training parameters using the options `--optimize` and `--optimize_out`. Functionality as previous command `logo` can be obtained with options `--groups` and `--rank`.
- `cccv`: roughly equivalent to previous command, same extras as `train`.
- `plot`: takes currently only the subcommand `cccv` which allows plotting of one or multiple CCCV results files in PhenDB format.

Click includes facilities for testing CLI components but as of now, the CLI is not tested (and is excluded from codecov).

Other minor changes:
- moved serialization functionality from utils to io
- bumped version
- added totally sick ascii art

### CODE REVIEW

File: pica/cli/generic_func.py
Comment (VarIr) [inline pica/cli/generic_func.py]: I'd rather have the first sentence similar to the second one, e.g. "Estimate model performance by cross-validation" (if that is what the function does).

File: pica/cli/generic_func.py
Comment (VarIr) [inline pica/cli/generic_func.py]: It seems the loaded params have precedence here (as in: if there are identical keys, `loaded_params` will overwrite those from `kwargs`), right?

File: pica/cli/generic_func.py
Comment (VarIr) [inline pica/cli/generic_func.py]: As in the other `generic_` functions, I'd prefer it like "Train some model"

File: pica/cli/generic_func.py
Comment (VarIr) [inline pica/cli/generic_func.py]: The same here. Though if you prefer it like it is now, that's also fine. Should at least be consistent style (which it is).

Also, please don't use the abbrev in the docstring, so that we can immediately what CCCV stands for in this place.

File: pica/cli/main.py
Comment (VarIr) [inline pica/cli/main.py]: Don't know why, but on github this is slightly distorted, while it looks fine if I copy it into gedit.
Otherwise, +1 for ANSI shadow :)

File: pica/cli/plot.py
Comment (VarIr) [inline pica/cli/plot.py]: Perhaps simply: "Plot results. [...]"

File: pica/cli/predict.py
Comment (VarIr) [inline pica/cli/predict.py]: Why not use print here?

File: pica/io/flat.py
Comment (VarIr) [inline pica/io/flat.py]: `NumpyEncoder` would be more articulate (while slightly longer)

File: pica/ml/classifiers/xgbm.py
Comment (VarIr) [inline pica/ml/classifiers/xgbm.py]: Why explicitly cast to int? Should be integer anyway.

File: pica/ml/classifiers/xgbm.py
Comment (VarIr) [inline pica/ml/classifiers/xgbm.py]: Here we have a mixture of Python lists and Numpy arrays. This does not matter as long as we only ever iterate over these structures. I'm only slightly worried that we at some point may do `gamma * 2` and obtain something rather surprising.

File: tests/test_data/Sulfate_reducer_xgb_cccv.json
Comment (VarIr) [inline tests/test_data/Sulfate_reducer_xgb_cccv.json]: I'm not so familiar with JSON, but POSIX standard would be to have a newline at EOF. I assume the JSON files are read/written by a library that just does not do that? Then it would probably not matter.

File: tests/test_data/svm_params.json
Comment (VarIr) [inline tests/test_data/svm_params.json]: Again, newline at EOF?

File: tests/test_data/xgb_params.json
Comment (VarIr) [inline tests/test_data/xgb_params.json]: Again, newline at EOF?

File: tests/test_data/Sulfate_reducer_svm_cccv.json
Comment (VarIr) [inline tests/test_data/Sulfate_reducer_svm_cccv.json]: Again, newline at EOF?

**Review by VarIr — COMMENTED** [review]
I did not know about click, but seems like a really cool package!
As usual, I have a few stylistic remarks, but in general this LGTM.
However, I just had a look at the code. Could you perhaps try to actually run the new cli, @phyden?

**Comment by VarIr** [comment]
Btw, while high code cov is a nice thing to have, I would not enforce it. Even 100% is no guarantee for error-free code, and code with incomplete coverage can be highly reliable (e.g. SciPy has 87% cov for Python code and only 45% for complied code). Still, some mechanism to ensure a functioning cli is a must at least for new releases on PyPI etc. Could be a handful of manual checks though.

File: tests/test_data/Sulfate_reducer_xgb_cccv.json
Comment (LokiLuciferase) [reply tests/test_data/Sulfate_reducer_xgb_cccv.json]: yeah, written and read with `json`. But inserting a newline won't hurt, you are right.

File: pica/cli/generic_func.py
Comment (LokiLuciferase) [reply pica/cli/generic_func.py]: yes, correct.

File: pica/cli/main.py
Comment (LokiLuciferase) [reply pica/cli/main.py]: think we can all agree this is the most significant improvement to trex since its inception.

File: pica/cli/predict.py
Comment (LokiLuciferase) [reply pica/cli/predict.py]: mumble mumble circumvent stdout buffer mumble mumble

...no idea in all honesty. I copied the code from the old predict function :)

**Comment by VarIr** [comment]
We have quite a number of flake8 warnings. I'll fix them (or set flake8 to ignore, in case of whitespace before a colon). Can I push to your feature branch? Or do I have to create a PR onto your PR?

**Comment by LokiLuciferase** [comment]
By all means, please do. Also feel free to implement any changes discussed if you have time, I probably won't until weekend at earliest.  😁  Do you have push access to my fork?

**Comment by VarIr** [comment]
Will try! I also don't have time actually, so I cannot guarantee for anything more.

**Comment by LokiLuciferase** [comment]
I found some time to apply the changes, thanks @VarIr ! By chance I found an issue with the CLI, guess it's really a good idea to test it through manually before merging and then write some tests soon.

The error: I forgot that some of the commands rely on a clean stdout stream, but the way I added my super rad ASCII art polluted it 🤷‍♂

File: pica/cli/cccv.py
Comment (VarIr) [inline pica/cli/cccv.py]: When I run `trex -h`, the description of CCCV is truncated like "Perform nested, stratified crossvalidation over a range of..." with the rest missing.

I think we should have a short descprition for `trex -h`, like "Simulated completeness/contamination crossvalidation". The rest does appear when calling `trex cccv -h`.

For example:
``` python3
@click.group()
def cccv():
    """
    Simulated completeness/contamination crossvalidation.

    Perform nested, stratified crossvalidation over a range of simulated completeness
    and contamination values for the training data.
    Optionally, save returned scores for each grid point.
    """
    pass
```

**Review by VarIr — COMMENTED** [review]
Thanks for the quick action! Just one more comment on the help texts of `click`, otherwise LGTM.

**Review by VarIr — APPROVED** [review]

**Review by VarIr — APPROVED** [review]

**Review by VarIr — APPROVED** [review]

**Pull request merged**
